File: src/compiler/lexer/token/StringToken.java
package compiler.lexer.token;

import compiler.parser.TokenEvaluator;
import compiler.parser.visitors.TokenVisitor;

public final class StringToken extends Token {
  private static final StringToken sentinel = new StringToken("");

  public StringToken(String str) {
    super(str.replace("\"", ""), 5);
  }

  public static StringToken getSentinel() {
    return sentinel;
  }

  @Override
  public <T> T accept(TokenVisitor<T> visitor) {
    return visitor.visit(this);
  }

  @Override
  public Object accept(TokenEvaluator visitor) throws Exception {
    return this.getValue();
  }
}

File: src/compiler/lexer/token/StringToken.kt
package compiler.lexer.token

import compiler.parser.TokenEvaluator
import compiler.parser.visitors.TokenVisitor

class StringToken(str: String) : Token(str.replace("\"", ""), 5) {
    override fun <T> accept(visitor: TokenVisitor<T>): T {
        return visitor.visit(this)
    }

    @Throws(Exception::class)
    override fun accept(visitor: TokenEvaluator): Any {
        return str
    }

    companion object {
        val sentinel = StringToken("")
    }
}
